### PR TITLE
[RDE operator] Implement CapiYaml retrieval from RDE

### PR DESCRIPTION
**Pull request by sahithi**
Implement CapiYaml retrieval from RDE; added Go types and CRDs.
kubectl apply is pending (r.client does not support apply, so looking for other alternatives)

Signed-off-by: sayloo <sahithi.ayloo@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/114)

### CODE REVIEW

**Review by Anirudh9794 — APPROVED** [review]
:lgtm: minor comments

> Reviewed 129 of 262 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/114#-:-N0cdzNcCIlXKC6pg5Le:b-kfw89e)* status: 129 of 262 files reviewed, 2 unresolved discussions (waiting on @Anirudh9794 and @sahithi)

___
*[`rdeprojector/api/v1alpha1/rdeprojector_types.go` line 38 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/114#-N0cb6os3mEOOkljDxLn:-N0cb6os3mEOOkljDxLo:bt9gv4s) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a54124e209cc3642387b2d9fb4d2d875288db4aa/rdeprojector/api/v1alpha1/rdeprojector_types.go#L38)):*
> ```Go
> 
> 	// +kubebuilder:validation:Required
> 	RDEId string `json:"rdeId,omitempty"`
> ```

I think omitempty should be removed if this field is a required property
___
*[`rdeprojector/controllers/rdeprojector_controller.go` line 73 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/114#-N0ccBc87S9YYFSMVUdI:-N0ccBc87S9YYFSMVUdJ:bo4tfo9) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a54124e209cc3642387b2d9fb4d2d875288db4aa/rdeprojector/controllers/rdeprojector_controller.go#L73)):*
> ```Go
> func (r *RDEProjectorReconciler) reconcileNormal(ctx context.Context, rdeProjector *capvcdv1alpha1.RDEProjector) (ctrl.Result, error) {
> 	log := ctrl.LoggerFrom(ctx)
> ```

We should have a logic to separate out userOrg and username from rdeProjector.Spec.UserCredentialsContext.Username

**Review by sahithi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/114)* status: 129 of 262 files reviewed, 2 unresolved discussions (waiting on @Anirudh9794)

___
*[`rdeprojector/api/v1alpha1/rdeprojector_types.go` line 38 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/114#-N0cb6os3mEOOkljDxLn:-N0fpOsK13Hw_Km8As0X:baw72ga) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a54124e209cc3642387b2d9fb4d2d875288db4aa/rdeprojector/api/v1alpha1/rdeprojector_types.go#L38)):*
<details><summary><i>Previously, Anirudh9794 wrote…</i></summary><blockquote>

I think omitempty should be removed if this field is a required property
</blockquote></details>

thanks, done.
___
*[`rdeprojector/controllers/rdeprojector_controller.go` line 73 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/114#-N0ccBc87S9YYFSMVUdI:-N0fp9ko6AofI5hcx4SH:b-cy2wd9) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a54124e209cc3642387b2d9fb4d2d875288db4aa/rdeprojector/controllers/rdeprojector_controller.go#L73)):*
<details><summary><i>Previously, Anirudh9794 wrote…</i></summary><blockquote>

We should have a logic to separate out userOrg and username from rdeProjector.Spec.UserCredentialsContext.Username
</blockquote></details>

Yes, there is a TODO right below to handle the sysadmin persona. I have requested Zichong to do this, as both are in the same capvcd repo.

File: rdeprojector/controllers/rdeprojector_controller.go [hunk 1]
```diff
@@ -60,3 +67,27 @@ func (r *RDEProjectorReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&capvcdv1alpha1.RDEProjector{}).
 		Complete(r)
 }
+
+func (r *RDEProjectorReconciler) reconcileNormal(ctx context.Context, rdeProjector *capvcdv1alpha1.RDEProjector) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	//TODO Ensure sysadmin persona works
+	workloadVCDClient, err := vcdsdk.NewVCDClientFromSecrets(rdeProjector.Spec.Site, rdeProjector.Spec.Org,
+		"", rdeProjector.Spec.Org, rdeProjector.Spec.UserCredentialsContext.Username,
```
Comment (arunmk) [inline rdeprojector/controllers/rdeprojector_controller.go:76]: We need the same userorg method here. Maybe that should become a common function in CPI common

File: rdeprojector/api/v1alpha1/rdeprojector_types.go [hunk 2]
```diff
@@ -23,19 +23,30 @@ import (
...
+
 // RDEProjectorSpec defines the desired state of RDEProjector
 type RDEProjectorSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
-	// Foo is an example field of RDEProjector. Edit rdeprojector_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// +kubebuilder:validation:Required
+	RDEId string `json:"rdeId,omitempty"`
```
Comment (arunmk) [inline rdeprojector/api/v1alpha1/rdeprojector_types.go:38]: The optional (`omitempty`) is because we can have it empty is it?

**Review by arunmk — APPROVED** [review]
:lgtm: thanks

**Pull request merged**